src/main/java/org/opensearch/knn/index/query/ExactSearcher.java

```diff
@@ -145,10 +142,43 @@ private KNNIterator getMatchedKNNIterator(
                 knnQuery.getQueryVector(),
                 (KNNFloatVectorValues) vectorValues,
                 spaceType,
-                knnQuery.getParentsFilter().getBitSet(leafReaderContext)
+                knnQuery.getParentsFilter().getBitSet(exactSearcherContext.getLeafReaderContext()),
+                quantizedQueryVector,
+                segmentLevelQuantizationInfo
             );
         }
 
-        return new FilteredIdsKNNIterator(matchedDocs, knnQuery.getQueryVector(), (KNNFloatVectorValues) vectorValues, spaceType);
+        return new FilteredIdsKNNIterator(
+            matchedDocs,
+            knnQuery.getQueryVector(),
+            (KNNFloatVectorValues) vectorValues,
+            spaceType,
+            quantizedQueryVector,
+            segmentLevelQuantizationInfo
+        );
+    }
+
+    /**
+     * Stores the context that is used to do the exact search. This class will help in reducing the explosion of attributes
+     * for doing exact search.
+     */
+    @Value
+    @Builder
+    public static class ExactSearcherContext {
+        /**
+         * controls whether we should use Quantized vectors during exact search or not. This is useful because when we do
+         * re-scoring we need to re-score using full precision vectors and not quantized vectors.
+         */
+        boolean useQuantizedVectorsForSearch;
+        LeafReaderContext leafReaderContext;
```

Comment:
[Not a blocker for release] nit: searchLeaf should be a function of LeafReaderContext and exactSearchParams, would have preferred it separate

Reply:
Sure. I don't see why we want to do it like that? Its just better to have 1 single class encapsulating the whole context of the search

Reply:
function signature can often be representations of what it does, in this case the function represents "search on a leaf(segment) given these params or a query".  It doesn't seem logical to club both data and params to search on that data especially when not needed.